core/src/middleware/mod.rs

```diff
@@ -314,13 +315,10 @@ pub trait RpcServiceT {
 	/// As a result, if you have custom logic for individual calls or notifications,
 	/// you must duplicate that implementation in this method or no middleware will be applied
 	/// for calls inside the batch.
-	fn batch<'a>(&self, requests: Batch<'a>) -> impl Future<Output = Result<Self::Response, Self::Error>> + Send + 'a;
+	fn batch<'a>(&self, requests: Batch<'a>) -> impl Future<Output = Self::BatchResponse> + Send + 'a;
 
 	/// Similar to `RpcServiceT::call` but processes a JSON-RPC notification.
-	fn notification<'a>(
-		&self,
-		n: Notification<'a>,
-	) -> impl Future<Output = Result<Self::Response, Self::Error>> + Send + 'a;
+	fn notification<'a>(&self, n: Notification<'a>) -> impl Future<Output = Self::NotificationResponse> + Send + 'a;
 }
 
 /// Interface for types that can be serialized into JSON.
```

Comment:
Nit: I'd be tempted to put this with the loggerm iddleware since that's where it's used, and then have the impls on specific types right next to it. That'll make it clear what the logger middleware will work with right off the bat :)